tools/kicad/kicad_rs/src/parser.rs

```diff
@@ -5,38 +5,92 @@ use std::path::Path;
 use crate::error::{errorf, DynamicResult};
 use crate::types::*;
 
-impl Schematic {
-    pub fn parse_id(path: &Path, id: String) -> DynamicResult<Schematic> {
-        let file = SchematicFile::load(path)?;
-        parse_schematic(&file, id)
-    }
-
-    pub fn parse(path: &Path) -> DynamicResult<Schematic> {
-        Self::parse_id(path, String::new())
-    }
+// SchematicTree keeps track of all kicad_parse_gen
+// Schematics in a hierarchical schematic configuration
+#[derive(Debug)]
+pub struct SchematicTree {
+    schematic: kicad_schematic::Schematic,
+    sub_schematics: HashMap<String, SchematicTree>,
 }
 
-pub struct SchematicFile<'a> {
-    path: &'a Path,
-    content: kicad_schematic::Schematic,
-}
+impl SchematicTree {
+    // Load a hierarchical SchematicTree from the given base schematic path
+    pub fn load(path: &Path) -> DynamicResult<Self> {
+        let mut sub_schematics = HashMap::new();
+        let schematic = kicad_schematic::parse_file(path)?;
+        for sub_sheet in schematic.sheets.iter() {
+            let filename = kicad_schematic::filename_for_sheet(&schematic, sub_sheet)?;
+            sub_schematics.insert(sub_sheet.name.clone(), SchematicTree::load(&filename)?);
+        }
 
-impl<'a> SchematicFile<'a> {
-    pub fn load(path: &'a Path) -> DynamicResult<Self> {
         Ok(Self {
-            path,
-            // Read the schematic contents using kicad_parse_gen
-            content: kicad_parse_gen::read_schematic(path)?,
+            schematic,
+            sub_schematics,
         })
     }
+
+    // Parse the SchematicTree into our own nested Schematic struct
+    pub fn parse(&self) -> DynamicResult<Schematic> {
+        parse_schematic(self, String::new())
+    }
+
+    // Update the components in the kicad_parse_gen Schematic tree using the given
+    // nested Schematic struct (copy values from Attributes to ComponentFields)
+    pub fn update(&mut self, schematic: &Schematic) -> DynamicResult<()> {
+        // Update the fields of all components in this schematic
+        for component in schematic.components.iter() {
+            self.schematic
+                .modify_component(&component.labels.reference, |c| {
+                    for attribute in component.attributes.iter() {
+                        let name = attribute.name.as_str().or_default("Value");
+                        c.update_field(name, &attribute.value.to_string());
+                        c.update_field(
+                            &format!("{}{}", name, "_expr"),
```

Comment:
Why do we write back the expression? Isn't the desired truth of that always set by the user in the kicad file?

Reply:
Indeed it is, but serialization of field values with quotes is broken in `kicad_parse_gen` (like the TODO I've added below that line states). This means that parsing and serializing back an Eeschema file with quotes while doing no modifications isn't reproducible in the first place, it will mess up escaping of the quotes. Since the whole file needs to be serialized at once (and not just some individual fields), I need to work around the bugs in serializing the expression as well, even though it is read-only internally.

Hardcoding `_expr` as the suffix isn't the most elegant solution, but it'll do for now IMO before I get around to fixing `kicad_parse_gen`. Then we can completely get rid of the `escape` and `unescape` functions here.